Add WKWebView SPI to fetch and restore Session Storage data

RupinMittal opened the pull request:
#### f52abbbacd4b9fd731e8a250e5e8c6379add50a5
<pre>
Add WKWebView SPI to fetch and restore Session Storage data
<a href="https://bugs.webkit.org/show_bug.cgi?id=283503">https://bugs.webkit.org/show_bug.cgi?id=283503</a>
<a href="https://rdar.apple.com/140356755">rdar://140356755</a>

Reviewed by Sihui Liu and Wenson Hsieh.

Some users may want Session Storage to be restored after a software
update. So we want to allow clients (Safari) to fetch this storage
and then later restore it.

The restoration process will have two parts:
1. Client fetches the session storage data from WebKit and holds onto it.
2. Client gives back the data to WebKit to restore it.

This patch adds SPI for fetch and restore--and will eventually go through
API review.

The SPI works for both ephemeral and persistent data stores, so there is
an API test for each. The SPI also works for both first party storage and
for third party storage, so there are API tests for each.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::fetchSessionStorage):
(WebKit::NetworkProcess::restoreSessionStorage):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
(WebKit::LocalStorageManager::setStorageMap):
(WebKit::LocalStorageManager::populateStorageArea): Deleted.
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::fetchSessionStorageForWebPage):
(WebKit::NetworkStorageManager::restoreSessionStorageForWebPage):
(WebKit::NetworkStorageManager::restoreLocalStorage):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/SessionStorageManager.cpp:
(WebKit::SessionStorageManager::fetchStorageMap):
(WebKit::SessionStorageManager::setStorageMap):
* Source/WebKit/NetworkProcess/storage/SessionStorageManager.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _fetchDataOfTypes:completionHandler:]):
(-[WKWebView _restoreData:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _fetchDataOfTypes:completionHandler:]):
(-[WKWebsiteDataStore _restoreData:completionHandler:]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::fetchSessionStorage):
(WebKit::WebPageProxy::restoreSessionStorage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchLocalStorage):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
(TEST(WebKit, RestoreLocalStorageFromPersistentDataStore)):
(TEST(WebKit, RestoreLocalStorageFromEphemeralDataStore)):
(RestoreLocalStorageFromPersistentDataStoreThirdPartyIFrame)):
(RestoreLocalStorageFromEphemeralDataStoreThirdPartyIFrame)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreSessionStorage.mm: Copied froTools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm.
(testRestoreSessionStorage):
(TEST(WebKit, RestoreSessionStorageFromPersistentDataStore)):
(TEST(WebKit, RestoreSessionStorageFromEphemeralDataStore)):
(-[RestoreSessionStorageMessageHandler userContentController:didReceiveScriptMessage:]):
(postMessage):
(item):
(RestoreSessionStorageFromPersistentDataStoreThirdPartyIFrame)):
(RestoreSessionStorageFromEphemeralDataStoreThirdPartyIFrame)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextWidth.mm:

Canonical link: <a href="https://commits.webkit.org/287729@main">https://commits.webkit.org/287729@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c5ad81493722270f9499419fa89459627f3929a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80614 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59621 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34548 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85138 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31596 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82725 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68682 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7927 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62966 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20765 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83683 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/53085 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73386 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43271 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50373 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27545 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30055 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71520 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28068 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86571 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7840 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5526 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71260 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8015 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69222 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70500 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14511 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13460 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12493 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7802 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13321 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7641 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11160 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9446 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -226,6 +226,15 @@ for this property.
+  @param dataTypes The additional types of data to fetch
+  @param completionHandler A block to invoke when the data has been fetched.
+*/
+- (void)fetchSessionStateAndAdditionalData:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler;
```
this should start with a _ since it's SPI for now

RupinMittal replied:
Fixed!

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -226,6 +226,15 @@ for this property.
 - (void)_frameTrees:(void (^)(NSSet<_WKFrameTreeNode *> *))completionHandler WK_API_AVAILABLE(macos(14.0), ios(17.0));
 
+/*! @constant WKWebViewSessionStateTypeSessionStorage Session Storage. */
+WK_EXTERN NSString * const WKWebViewSessionStateTypeSessionStorage WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
having constants in new SPI/API is generally discouraged I think. Can this be an NS_ENUM instead?

RupinMittal replied:
I believe we can't have an NSSet of NS_ENUMs.

cdumez replied:
Why don't we simply reuse `WKWebsiteDataTypeSessionStorage` like in the WKWebsiteDataStore API?

RupinMittal replied:
@rr-codes I tried NS_ENUM and NS_OPTIONS but it seems neither can be held inside a NSSet. I see other APIs using NSString so that's what I have here at the moment. Do you have any other ideas of what to use or is NSString the way to go?

whsieh commented on Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:
```diff
@@ -477,28 +480,31 @@ static RetainPtr<CFDictionaryRef> encodeSessionHistory(const BackForwardListStat
+    //     sessionStorageDictionary = encodeSessionStorage(*sessionStorage);
+
+    auto stateDictionary = CFDictionaryCreateMutable(
+        kCFAllocatorDefault, 0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
```
This looks like a memory leak — you'll want to use `adoptCF` here.
```suggestion
    auto stateDictionary = adoptCF(CFDictionaryCreateMutable(
        kCFAllocatorDefault, 0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks));
```

RupinMittal replied:
Ah, thank you for catching that!

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -601,6 +601,11 @@ typedef NS_OPTIONS(NSUInteger, WKDisplayCaptureSurfaces) {
+/*! @constant WKWebViewDataTypeSessionStorage */
+WK_EXTERN NSString * const WKWebViewDataTypeSessionStorage WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionOS(WK_XROS_TBA));
+
+- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionOS(WK_XROS_TBA));
```
I am not sure why this takes `type` as parameters;  the other types are not stored by view so I don't think we expect them to be fetched with this SPI. `_fetchSessionStorageData` seems more reasonable.

RupinMittal replied:
Oh I see. I was trying to make it generic in case we add more types later. But that makes sense--if the other types aren't even stored on the view, we won't use this SPI for them. Updated!

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -830,6 +830,31 @@ void NetworkStorageManager::cloneSessionStorageNamespace(StorageNamespaceIdentif
+
+        HashMap<WebCore::ClientOrigin, HashMap<String, String>> sessionStorageMap;
+
+        for (auto& origin : getAllOrigins()) {
```
This should be `m_originStorageManagers`. SessionStorage data is stored in memory, so we only need to search in existing `OriginStorageManager`s

szewai replied:
This should be `m_originStorageManagers`. SessionStorage data is stored in memory, so we only need to search existing `OriginStorageManager`s.

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -830,6 +830,31 @@ void NetworkStorageManager::cloneSessionStorageNamespace(StorageNamespaceIdentif
+        HashMap<WebCore::ClientOrigin, HashMap<String, String>> sessionStorageMap;
+
+        for (auto& origin : getAllOrigins()) {
+            auto sessionStorageManager = originStorageManager(origin, ShouldWriteOriginFile::No).sessionStorageManager((*m_storageAreaRegistry));
```
Ditto, this should be `existingSessionStorageManager`

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreSessionStorage.mm:
```diff
@@ -0,0 +1,188 @@
+    __block RetainPtr<NSData> sessionStorageData;
+    [webView _fetchDataOfTypes:set.get() completionHandler:^(NSData *data) {
+        EXPECT_NOT_NULL(data);
+        sessionStorageData = data;
```
`sessionStorageData` is unused

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreSessionStorage.mm:
```diff
@@ -0,0 +1,188 @@
+
+TEST(WebKit, FetchSessionStorageFromPersistentDataStore)
+{
+    testFetchSessionStorage(adoptNS([WKWebsiteDataStore defaultDataStore]));
```
`adoptNS` should be be removed

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreSessionStorage.mm:
```diff
@@ -0,0 +1,188 @@
+
+TEST(WebKit, FetchSessionStorageFromEphemeralDataStore)
+{
+    testFetchSessionStorage(adoptNS([WKWebsiteDataStore nonPersistentDataStore]));
```
Ditto

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreSessionStorage.mm:
```diff
@@ -0,0 +1,188 @@
+
+TEST(WebKit, FetchSessionStorageFromPersistentDataStoreThirdPartyIFrame)
+{
+    testFetchSessionStorageThirdPartyIFrame(adoptNS([WKWebsiteDataStore defaultDataStore]));
```
Ditto

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreSessionStorage.mm:
```diff
@@ -0,0 +1,188 @@
+
+TEST(WebKit, FetchSessionStorageFromEphemeralDataStoreThirdPartyIFrame)
+{
+    testFetchSessionStorageThirdPartyIFrame(adoptNS([WKWebsiteDataStore nonPersistentDataStore]));
```
Ditto

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreSessionStorage.mm:
```diff
@@ -0,0 +1,188 @@
+    __block RetainPtr<NSData> sessionStorageData;
+    [webView _fetchDataOfTypes:set.get() completionHandler:^(NSData *data) {
+        EXPECT_NOT_NULL(data);
+        sessionStorageData = data;
```
`sessionStorageData` is not used

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -830,6 +830,33 @@ void NetworkStorageManager::cloneSessionStorageNamespace(StorageNamespaceIdentif
+            if (!sessionStorageManager)
+                continue;
+
+            auto storageNameSpaceIdentifier = ObjectIdentifier<StorageNamespaceIdentifierType>(pageIdentifier.toUInt64());
```
`StorageNamespaceIdentifier storageNameSpaceIdentifier { pageIdentifier.toUInt64() };`

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -601,6 +601,8 @@ typedef NS_OPTIONS(NSUInteger, WKDisplayCaptureSurfaces) {
 // This property is KVO compliant.
 @property (nonatomic, readonly) _WKWebProcessState _webProcessState WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
 
+- (void)_fetchSessionStorageData:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionOS(WK_XROS_TBA));
```
What happened to the generic API that takes in a NSSet of types, like we did for the WKWebsiteDataStore?

This doesn't look like this can become API?

RupinMittal replied:
Just brought this back as discussed.

szewai reviewed (commented):
The PR title is wrong, it's `WKWebView` SPI

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -830,6 +830,33 @@ void NetworkStorageManager::cloneSessionStorageNamespace(StorageNamespaceIdentif
+            if (!sessionStorageManager)
+                continue;
+
+            StorageNamespaceIdentifier storageNameSpaceIdentifier { pageIdentifier.toUInt64() };
```
This can be moved above the for loop -- you don't need to create `StorageNamespaceIdentifier` for each origin.

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -5244,6 +5246,62 @@ - (_WKWebProcessState)_webProcessState
+
+    if (dataTypes & _WKWebViewDataTypeSessionStorage) {
+        RefPtr page = [self _protectedPage];
+        page->fetchSessionStorage([callbackAggregator, protectedPage = page, data](HashMap<WebCore::ClientOrigin, HashMap<String, String>> sessionStorage) {
```
`auto&& sessionStorage`

szewai commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15648,6 +15648,11 @@ FloatPoint WebPageProxy::mainFrameScrollPosition() const
 
+void WebPageProxy::fetchSessionStorage(CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&)>&& completionHandler)
+{
+    protectedWebsiteDataStore()->protectedNetworkProcess()->sendWithAsyncReply(Messages::NetworkProcess::FetchSessionStorage(sessionID(), identifier()), WTFMove(completionHandler));
```
`if (RefPtr networkProcess = websiteDataStore().networkProcessIfExists())`
If network process does not exist yet, we don't need to launch one to fetch data

RupinMittal replied:
Fixed!

szewai reviewed (approved):
LGTM with nits

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -601,6 +601,12 @@ typedef NS_OPTIONS(NSUInteger, WKDisplayCaptureSurfaces) {
 
+typedef NS_OPTIONS(NSUInteger, _WKWebViewDataType) {
+    _WKWebViewDataTypeSessionStorage = 0
+} WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionOS(WK_XROS_TBA));
```
Lowercase `visionos` here.

```suggestion
} WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
```

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -601,6 +601,12 @@ typedef NS_OPTIONS(NSUInteger, WKDisplayCaptureSurfaces) {
+    _WKWebViewDataTypeSessionStorage = 0
+} WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionOS(WK_XROS_TBA));
+
+- (void)_fetchDataOfTypes:(_WKWebViewDataType)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionOS(WK_XROS_TBA));
```
Here too

```suggestion
- (void)_fetchDataOfTypes:(_WKWebViewDataType)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
```

RupinMittal replied:
Fixed!

RupinMittal committed:
Add WKWebView SPI to fetch and restore Session Storage data
https://bugs.webkit.org/show_bug.cgi?id=283503
rdar://140356755

Reviewed by Sihui Liu and Wenson Hsieh.

Some users may want Session Storage to be restored after a software
update. So we want to allow clients (Safari) to fetch this storage
and then later restore it.

The restoration process will have two parts:
1. Client fetches the session storage data from WebKit and holds onto it.
2. Client gives back the data to WebKit to restore it.

This patch adds SPI for fetch and restore--and will eventually go through
API review.

The SPI works for both ephemeral and persistent data stores, so there is
an API test for each. The SPI also works for both first party storage and
for third party storage, so there are API tests for each.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::fetchSessionStorage):
(WebKit::NetworkProcess::restoreSessionStorage):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
(WebKit::LocalStorageManager::setStorageMap):
(WebKit::LocalStorageManager::populateStorageArea): Deleted.
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::fetchSessionStorageForWebPage):
(WebKit::NetworkStorageManager::restoreSessionStorageForWebPage):
(WebKit::NetworkStorageManager::restoreLocalStorage):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/SessionStorageManager.cpp:
(WebKit::SessionStorageManager::fetchStorageMap):
(WebKit::SessionStorageManager::setStorageMap):
* Source/WebKit/NetworkProcess/storage/SessionStorageManager.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _fetchDataOfTypes:completionHandler:]):
(-[WKWebView _restoreData:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _fetchDataOfTypes:completionHandler:]):
(-[WKWebsiteDataStore _restoreData:completionHandler:]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::fetchSessionStorage):
(WebKit::WebPageProxy::restoreSessionStorage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchLocalStorage):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
(TEST(WebKit, RestoreLocalStorageFromPersistentDataStore)):
(TEST(WebKit, RestoreLocalStorageFromEphemeralDataStore)):
(RestoreLocalStorageFromPersistentDataStoreThirdPartyIFrame)):
(RestoreLocalStorageFromEphemeralDataStoreThirdPartyIFrame)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreSessionStorage.mm: Copied froTools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm.
(testRestoreSessionStorage):
(TEST(WebKit, RestoreSessionStorageFromPersistentDataStore)):
(TEST(WebKit, RestoreSessionStorageFromEphemeralDataStore)):
(-[RestoreSessionStorageMessageHandler userContentController:didReceiveScriptMessage:]):
(postMessage):
(item):
(RestoreSessionStorageFromPersistentDataStoreThirdPartyIFrame)):
(RestoreSessionStorageFromEphemeralDataStoreThirdPartyIFrame)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextWidth.mm:

Canonical link: https://commits.webkit.org/287729@main
